### PR TITLE
editor: Fix hover hiding delay when `symbol_range` is zero-width

### LINKED ISSUE #46841: Diagnostic hover popover disappears when moving mouse cursor to the popover
### Reproduction steps

1. Add this to your `pyproject.toml`:

```toml
[tool.ruff.lint]
extend-select = ["I"]
```

2. Save this Python snippet to a file:

```python
#! /usr/bin/env -S uv run --script
#
# /// script
# dependencies = []
# ///

from datetime import UTC
import datetime as dt
from typing import TYPE_CHECKING, Any, Self

print(dt, UTC, TYPE_CHECKING, Any, Self)
```

3. Hover the mouse cursor over the end of line 9. The diagnostic hover popup appears above_line 7.
4. Move your mouse cursor to the hover popup.

### Current vs. Expected behavior

Ruff warns about [unsorted-imports](https://docs.astral.sh/ruff/rules/unsorted-imports/) on lines 7-9. Line 9 and the hover popup are both long. But lines 7-8 are short, so there is some empty space, which Zed doesn't consider to be part of the warning.

If you put the mouse cursor near the end of line 9 to see the hover popup, you will probably move it straight upward to the hover popup, which causes the popup to disappear.

https://github.com/user-attachments/assets/310e0dbf-6282-4f7b-bcf9-8ca7486c7baa

### Zed version and system specs

Zed: v0.219.4+stable.119.5595f7ce23d51a58eb66dd025a487bbd6c75dad1 (Zed) 
OS: macOS 15.7.3
Memory: 16 GiB
Architecture: x86_64

### PR DESCRIPTION
**Pull request by daydalek**
Previously, for language servers that omit the range field, every slight mouse movement triggered a new hover request and dismissed the old popover, making it impossible to interact with. This PR fixes it by introducing a small equivalence window for zero-width hover targets (start == end), so nearby offsets are treated as the same hover target instead of immediate target switches.

Closes #46841 

Release Notes:

- Fixed hover popover flickering and interaction issues with language servers that omit the range field in hover responses.

### CODE REVIEW

**Comment by daydalek** [comment]
I initially tried to balance the experience by keeping the old hover visible while a new hover request was in flight. In practice, that approach regressed responsiveness: when moving across hoverable code, the popover keeps shown instead of following the usual hide/show timing behavior. That made normal hover interactions feel less responsive than before. I guess the simplest way would be handling only zero-width cases directly by introducing an offset without changing the lifecycle behavior previously introduced, but the threshold is still heuristic: we do not yet know what offset value is best across languages, fonts and input patterns. It is also unclear whether this should remain an internal constant or become a user-facing setting. Please feel free to comment candidly on the offset approach: whether it’s the right fix, what threshold makes sense, and whether this should be configurable or kept as a constant.

**Comment by daydalek** [comment]
@codex review

File: crates/editor/src/hover_popover.rs [hunk 1]
```diff
@@ -300,7 +302,6 @@ fn show_hover(
     if !ignore_timeout {
         if same_info_hover(editor, &snapshot, anchor)
             || same_diagnostic_hover(editor, &snapshot, anchor)
-            || editor.hover_state.diagnostic_popover.is_some()
         {
```
[the remark this answers is not in the document]
Comment (daydalek) [reply crates/editor/src/hover_popover.rs:305]: I have now put it back.

**Review by SomeoneToIgnore — COMMENTED** [review]
Sorry for the late review, was on a hiatus most of the time.

Coming back and wondering about the heuristics: would picking a `surrounding_word`'s coordinate be better here instead of `anchor..anchor`?

The issues I see from the quick glance: 
* seems that the bug is still there for the last excerpt's last offset, because `hover_excerpt_range.contains(&hover_offsets.start)` is exclusive at the end
* +-4 byte offset manipulation is somewhat dangerous in the world of Unicode and things like √√√√√√√

**Comment by daydalek** [comment]
> Sorry for the late review, was on a hiatus most of the time.
> 
> Coming back and wondering about the heuristics: would picking a `surrounding_word`'s coordinate be better here instead of `anchor..anchor`?
> 
> The issues I see from the quick glance:
> 
> * seems that the bug is still there for the last excerpt's last offset, because `hover_excerpt_range.contains(&hover_offsets.start)` is exclusive at the end
> * +-4 byte offset manipulation is somewhat dangerous in the world of Unicode and things like √√√√√√√

Thank you for the review—both points make sense. I’m going to rework this using surrounding_word and properly handle the excerpt boundary cases. I’ll follow up once the implementation and tests are ready.

**Comment by SomeoneToIgnore** [comment]
Great, thanks for the acknowledgement — will close the PR for now as it's been a bit over 3 weeks with not actions, and we want to keep the PR queue cleaner.

Happy to review that later.
